oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,270 @@
+
+Alternatives Considered
+=======================
+
+We chose to avoid putting the locale in the domain name itself, to leave that available for other uses. We considered but rejected putting it in as a URL parameter; that can be harder to parse out, doesn't give an SEO benefit, and doesn't give us automatic separation of analytics by locale. Using the first component of the path ensures that if users send a link or bookmark a page, their recipients will see the exact content that they intended to share.
+
+***********************
+Translations Management
+***********************
+
+Message Ids
+===========
+
+Message ids, or message keys, uniquely identify a string to be internationalized. They should consist of one or more words to namespace the message, plus one or more words to identify the message. All words will be lowercase and separated by periods. Here are examples of message ids used in a header and in a login modal, respectively:
+
+  | ``header.search``
+  | ``login.no.such.password``
+  | ``login.privacy.policy``
+
+The part of the id used for namespacing is purely to reduce collisions and help human readability; it's not intended to be programmatically parsable. That said, a helpful convention is to make it the name of your application or repo. Note that if you end up with a duplicate message id in the same translation project in Transifex, and they have different content, one of them will "win" when you download your translated strings, leading to an out-of-place string in your application.
+
+Alternatives Considered
+-----------------------
+
+Previous i18n implementations have sometimes used the English text as the message id. While this saves developers a step in their work, it has the drawback that a change in the English copy invalidates all existing translations, even if they would not otherwise have needed to change. (For example, capitalizing a noun in English has no impact on the German translation, because nouns are capitalized in German anyway. Or an English change might be correcting a typo that was correct in the translations.) Do not use English copy as the message id.
+
+Cross-Site Scripting Vulnerability
+==================================
+
+Because translated strings often contain placeholders that will be filled in at render time, possibly using user input, localization is a potential opening for cross-site scripting (XSS) attacks. The ``react-intl`` functions exposed by frontend-platform are XSS-safe.
+
+We explicitly don't export ``react-intl`` components that are not XSS safe, such as ``FormattedHTMLMessage``.  See `Don't Use FormattedHTMLMessage` <https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/preventing_xss/preventing_xss_in_react.html#don-t-use-formattedhtmlmessage>`_ for more information.
```

Comment:
After reviewing what they changed in more recent versions, I think this section is technically correct and a useful warning.

The change in more recent versions of react-intl is to deprecate FormattedHTMLMessage in favor of a new form of FormattedMessage which allows you to explicitly opt-in to HTML Parsing: https://formatjs.io/docs/react-intl/upgrade-guide-4x/#breaking-api-changes

After we upgrade react-intl this section will be _meaningless_ rather than _wrong_, which I think is acceptable until we come back and remove it. :)